### ISSUE TITLE
File Path

### ISSUE THREAD

**Issue by liamb8** [nestordemeure/GPTranslate#3]
How should I be putting the file path in gptranslate.py? I see that there's a variable data_folder but when changing the path there to the directory path that the file I'm trying to translate is in it doesn't work. This is the error that it's throwing when trying to run gptranslate.py UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4: character maps to < undefined >.

**Comment by nestordemeure** [comment]
Yes, by default you need to set `data_folder` (to the path to the directory containing the data) and/or `source_file` (to the actual file).
As a first test you might want to use a `.txt` file and convert a paragraph of one European language to another (French to English for example, two languages with a latin alphabet).

Your error looks like it found the file but its encoding is not recognized. If the first test works you might want to look into the encoding of your file and maybe reexporting or reencoding it into something like utf-8 (guessing that your source language does not use the latin alphabet: that should be fine as long as the encoding is recognized).

**Comment by liamb8** [comment]
I checked the file and it's in utf-8 and tried to also reencode it to utf-8 again but am still getting the same error message. I was able to do the test you mentioned translating from French to English. I'm attempting to translate from Korean to English though which is giving me this error.

**Comment by nestordemeure** [comment]
Okay, so it is definitively an encoding problem. I have done tests with Russian so i know non-latin alphabets should work.

You could try loading your file as raw text into Python and printing it to see if it works properly. It might be only a handful of chars causing the problem.

**Comment by liamb8** [comment]
Okay, I think the problem is with loading the file. I was able to get the Korean contents of the file printed out using:
`b = open(source_file, "r", encoding="utf-8").read()`
`print(b)`
The problem is I'm not sure how to reflect that with this `book = Book.load(source_file)` since it throws a TypeError if I add `"r", encoding="utf-8"`.

**Comment by nestordemeure** [comment]
I just added `encoding="utf-8"` to the code reading text files, give it a try and tell me if it fixes the problem!

**Comment by liamb8** [comment]
Yes, that worked! Works without any issues now. Is there a way to speed up the translation process if I'm translating a lot at once? Since I'm translating an entire Korean novel that has around 500 chapters. Also how do I change the model if I want to use GPT-4 since I have API access to it?

**Comment by nestordemeure** [comment]
Great!

To speed up the translation, you can pass the book as a single structured file (`epub` or `html`), it will then translate all chapters concurrently (if you have a folder of texts, it should be fairly straightforward to convert them into a rough html whose headings reflect the chapter structure).

I have not tested the code with gpt4 but it should run without problems. To do so, you need to pass `model="gpt-4"` to `ChatOpenAI` [here](https://github.com/nestordemeure/GPTranslate/blob/57cf9c5d48982771e3218a51a602b7e0d1817fe7/GPTranslate/translation/translator.py#LL62C9-L62C19). Note that it will make the translation significantly more expensive (translation can already be a lot of queries, I would run it with gpt3.5 first to get an idea of the price).

**Comment by liamb8** [comment]
Thanks, I did some testing earlier and saw that switching to GPT-4 was definitely not cheap (costed me $7 for one chapter). I converted the txt document to an html file but I'm running into some errors when it attempts to translate it. It seems that it's surpassing the amount of maximum context length for tokens. This is the error that it gave me openai.error.InvalidRequestError: This model's maximum context length is 4097 tokens. However, your messages resulted in 525372 tokens. Please reduce the length of the messages. I'm trying to translate 11 chapters of a Korean novel in one html document. It also didn't seem to recognize the language when I changed it to html format and said it was English. I ended up changing it to Korean for the language_source but I'm not sure if I'm getting this error because it's not reading the file correctly and recognizing the text.

It seems to work with the epub format though. So waiting on seeing the results with that. Also is there a way for the translation to be more accurate? Since when I translated 1 chapter in a file vs 11 in one file the translation with the 11 chapter one was worse than the one with just one chapter.

**Comment by nestordemeure** [comment]
Yes, I might be able to reduce token number / prices by something like 10% but GPT4 would still be really expensive.

You can specify the langage source manually [here](https://github.com/nestordemeure/GPTranslate/blob/57cf9c5d48982771e3218a51a602b7e0d1817fe7/gptranslate.py#LL7C1-L7C16) to avoid any risk of mistake.

That many tokens causing an overflow looks like it tried to translate a full chapter as a single block (not a good idea for a large chapter), using paragraphs markers should let it split the text into successive blocks (`<p>Text of a paragraph.</p>`), I suspect epub gives you that automatically (this is something I will fix if i come back to spending a significant amount of time on this project, I have now some good practice on working around overflow).

If the 11 chapter version did process the 11 chapters concurrently and **didn't cut into smaller units** (you should be able tot check that with the text displayed while translating) then you should be fine and the degradation of quality that you saw is purely bad luck (the translation is non deterministic so running it twice will result in slight variations).
The translation quality is impacted by the choice of model and amount of context already available to translate.
If translation quality is what matters most, there is a manual mode that will suggest several translation and have you pick one or write your own (but, obviously, it does not scale well to 500 chapters).

(Overall, you are running into the fact that this code works but is not as polished as a production ready piece of code should be.)
